lib/Frontend/Frontend.cpp

```diff
@@ -138,70 +158,136 @@ bool CompilerInstance::setup(const CompilerInvocation &Invok) {
-       i != e; ++i) {
-    bool hasError = setUpForFileAt(i);
-    if (hasError) {
+bool CompilerInstance::setUpForFile(StringRef fileName, bool isPrimary) {
+  if (fileName.empty())
+    return false;
+  // FIXME: Working with filenames is fragile, maybe use the real path
+  // or have some kind of FileManager.
+  using namespace llvm::sys::path;
+  if (Optional<unsigned> existingBufferID =
+          SourceMgr.getIDForBufferIdentifier(fileName)) {
+    if (isInSILMode() ||
+        (isInputSwift() && filename(fileName) == "main.swift")) {
+      assert(MainBufferID == NO_SUCH_BUFFER && "re-setting MainBufferID");
+      MainBufferID = existingBufferID.getValue();
+    }
+
+    if (isPrimary) {
+      assert(PrimaryBufferID == NO_SUCH_BUFFER && "re-setting PrimaryBufferID");
+      PrimaryBufferID = existingBufferID.getValue();
+    }
+
+    return false; // replaced by a memory buffer.
+  }
+
+  // Open the input file.
+  using FileOrError = llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>>;
+  FileOrError inputFileOrErr = llvm::MemoryBuffer::getFileOrSTDIN(fileName);
+  if (!inputFileOrErr) {
+    Diagnostics.diagnose(SourceLoc(), diag::error_open_input_file, fileName,
+                         inputFileOrErr.getError().message());
+    return true;
+  }
+
+  if (serialization::isSerializedAST(inputFileOrErr.get()->getBuffer())) {
+    llvm::SmallString<128> ModuleDocFilePath(fileName);
+    llvm::sys::path::replace_extension(ModuleDocFilePath,
+                                       SERIALIZED_MODULE_DOC_EXTENSION);
+    FileOrError moduleDocOrErr =
+        llvm::MemoryBuffer::getFileOrSTDIN(ModuleDocFilePath.str());
+    if (!moduleDocOrErr &&
+        moduleDocOrErr.getError() != std::errc::no_such_file_or_directory) {
+      Diagnostics.diagnose(SourceLoc(), diag::error_open_input_file, fileName,
+                           moduleDocOrErr.getError().message());
       return true;
     }
+    PartialModules.push_back(
+        {std::move(inputFileOrErr.get()),
+         moduleDocOrErr ? std::move(moduleDocOrErr.get()) : nullptr});
+    return false;
   }
+  // Transfer ownership of the MemoryBuffer to the SourceMgr.
+  unsigned bufferID =
+      SourceMgr.addNewSourceBuffer(std::move(inputFileOrErr.get()));
 
-  // Set the primary file to the code-completion point if one exists.
-  if (CodeCompletionBufferID.hasValue())
-    PrimaryBufferID = *CodeCompletionBufferID;
+  InputSourceCodeBufferIDs.push_back(bufferID);
 
-  if (MainMode && MainBufferID == NO_SUCH_BUFFER &&
-      InputSourceCodeBufferIDs.size() == 1)
-    MainBufferID = InputSourceCodeBufferIDs.front();
+  if (isInSILMode() || (isInputSwift() && filename(fileName) == "main.swift")) {
+    assert(MainBufferID == NO_SUCH_BUFFER && "re-setting MainBufferID");
+    MainBufferID = bufferID;
+  }
+
+  if (isPrimary) {
+    assert(PrimaryBufferID == NO_SUCH_BUFFER && "re-setting PrimaryBufferID");
+    PrimaryBufferID = bufferID;
+  }
```

Comment:
These two `if`s seem like a possible target for (later) refactoring; we've seen them at least 3 times by now.

Reply:
Yes, they will be in an upcoming PR.